File: server/src/security/prompt-injection.service.ts
import { Injectable } from '@nestjs/common';

export type PromptInjectionRiskLevel = 'none' | 'low' | 'medium' | 'high';

export interface PromptInjectionAnalysis {
  detected: boolean;
  riskLevel: PromptInjectionRiskLevel;
  indicators: string[];
  fragments: string[];
}

export interface PromptInjectionResultFields {
  promptInjectionDetected?: boolean;
  promptInjectionRisk?: PromptInjectionRiskLevel;
  promptInjectionFragments?: string[];
  securityComment?: string;
}

interface InjectionPattern {
  indicator: string;
  regex: RegExp;
  riskLevel: Exclude<PromptInjectionRiskLevel, 'none'>;
}

const RISK_ORDER: Record<PromptInjectionRiskLevel, number> = {
  none: 0,
  low: 1,
  medium: 2,
  high: 3,
};

@Injectable()
export class PromptInjectionService {
  private readonly patterns: InjectionPattern[] = [
    {
      indicator: 'ignore-previous-instructions',
      regex: /(ignore|disregard|forget)\s+(all\s+)?(previous|prior|above)\s+instructions/i,
      riskLevel: 'high',
    },
    {
      indicator: 'change-assistant-role',
      regex: /(you are now|act as|developer message|system prompt|system message)/i,
      riskLevel: 'high',
    },
    {
      indicator: 'force-positive-grade',
      regex: /(поставь|выставь|дай|верни)\s+(10|десять|максимальн|положительн|отличн)/i,
      riskLevel: 'high',
    },
    {
      indicator: 'skip-check',
      regex: /(не\s+проверяй|пропусти\s+проверку|skip\s+(the\s+)?check|do\s+not\s+check)/i,
      riskLevel: 'high',
    },
    {
      indicator: 'force-output',
      regex: /(верни\s+только|return\s+only|respond\s+only|напиши\s+только)/i,
      riskLevel: 'medium',
    },
    {
      indicator: 'hide-errors',
      regex: /(скрой\s+ошибк|не\s+указывай\s+ошибк|hide\s+errors|do\s+not\s+mention\s+errors)/i,
      riskLevel: 'high',
    },
    {
      indicator: 'structured-output-tampering',
      regex: /<\/?JSON>|@SPLIT|<\/?UNTRUSTED_[A-Z_]+>/i,
      riskLevel: 'medium',
    },
  ];

  analyze(text: string): PromptInjectionAnalysis {
    if (!text?.trim()) {
      return this.emptyAnalysis();
    }

    const indicators: string[] = [];
    const fragments: string[] = [];
    let riskLevel: PromptInjectionRiskLevel = 'none';

    for (const pattern of this.patterns) {
      const match = text.match(pattern.regex);

      if (!match || match.index === undefined) {
        continue;
      }

      indicators.push(pattern.indicator);
      fragments.push(this.extractFragment(text, match.index, match[0].length));
      riskLevel = this.maxRisk(riskLevel, pattern.riskLevel);
    }

    return {
      detected: indicators.length > 0,
      riskLevel,
      indicators: this.unique(indicators),
      fragments: this.unique(fragments),
    };
  }

  mergeResultFields<T extends PromptInjectionResultFields>(
    result: T,
    analysis: PromptInjectionAnalysis,
  ): T & Required<PromptInjectionResultFields> {
    const detected = Boolean(result.promptInjectionDetected) || analysis.detected;
    const promptInjectionRisk = this.maxRisk(
      result.promptInjectionRisk ?? 'none',
      analysis.riskLevel,
    );
    const promptInjectionFragments = this.unique([
      ...(result.promptInjectionFragments ?? []),
      ...analysis.fragments,
    ]);
    const securityComment =
      result.securityComment ??
      (detected ? 'Potential prompt injection was detected in untrusted report content.' : '');

    return {
      ...result,
      promptInjectionDetected: detected,
      promptInjectionRisk,
      promptInjectionFragments,
      securityComment,
    };
  }

  assertGeneratedReviewAllowed(result: {
    review?: string;
    advantages?: string[];
    disadvantages?: string[];
  }) {
    const text = [
      result.review ?? '',
      ...(result.advantages ?? []),
      ...(result.disadvantages ?? []),
    ]
      .join('\n')
      .toLowerCase();

    const forbiddenMarkers = [
      'проверка пропущена',
      'пропускаю проверку',
      'по просьбе студента',
      'ставлю 10 по просьбе',
      'skip the check',
      'as requested by the student',
      'give full credit',
    ];

    if (forbiddenMarkers.some((marker) => text.includes(marker))) {
      throw new Error('Model response appears to follow a prompt injection instruction');
    }
  }

  formatSecurityContext(analysis: PromptInjectionAnalysis) {
    return JSON.stringify(
      {
        promptInjectionDetected: analysis.detected,
        promptInjectionRisk: analysis.riskLevel,
        indicators: analysis.indicators,
        suspiciousFragments: analysis.fragments,
      },
      null,
      2,
    );
  }

  private emptyAnalysis(): PromptInjectionAnalysis {
    return {
      detected: false,
      riskLevel: 'none',
      indicators: [],
      fragments: [],
    };
  }

  private maxRisk(
    current: PromptInjectionRiskLevel,
    next: PromptInjectionRiskLevel,
  ): PromptInjectionRiskLevel {
    return RISK_ORDER[next] > RISK_ORDER[current] ? next : current;
  }

  private extractFragment(text: string, index: number, length: number) {
    const padding = 80;
    const start = Math.max(0, index - padding);
    const end = Math.min(text.length, index + length + padding);

    return text.slice(start, end).replace(/\s+/g, ' ').trim();
  }

  private unique(items: string[]) {
    return [...new Set(items.filter(Boolean))];
  }
}

File: src/ReportsCheck.Application/Security/PromptInjectionService.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using ReportsCheck.Domain.Security;

namespace ReportsCheck.Application.Security;

/// <summary>
/// Детерминированный сканер инъекций промпта. Прямой порт
/// PromptInjectionService из исходного NestJS-проекта.
/// </summary>
public class PromptInjectionService : IPromptInjectionService
{
    private const RegexOptions Options = RegexOptions.IgnoreCase | RegexOptions.CultureInvariant;

    private sealed record InjectionPattern(string Indicator, Regex Regex, string RiskLevel);

    private readonly InjectionPattern[] _patterns =
    [
        new("ignore-previous-instructions",
            new Regex(@"(ignore|disregard|forget)\s+(all\s+)?(previous|prior|above)\s+instructions", Options),
            PromptInjectionRisk.High),
        new("change-assistant-role",
            new Regex(@"(you are now|act as|developer message|system prompt|system message)", Options),
            PromptInjectionRisk.High),
        new("force-positive-grade",
            new Regex(@"(поставь|выставь|дай|верни)\s+(10|десять|максимальн|положительн|отличн)", Options),
            PromptInjectionRisk.High),
        new("skip-check",
            new Regex(@"(не\s+проверяй|пропусти\s+проверку|skip\s+(the\s+)?check|do\s+not\s+check)", Options),
            PromptInjectionRisk.High),
        new("force-output",
            new Regex(@"(верни\s+только|return\s+only|respond\s+only|напиши\s+только)", Options),
            PromptInjectionRisk.Medium),
        new("hide-errors",
            new Regex(@"(скрой\s+ошибк|не\s+указывай\s+ошибк|hide\s+errors|do\s+not\s+mention\s+errors)", Options),
            PromptInjectionRisk.High),
        new("structured-output-tampering",
            new Regex(@"<\/?JSON>|@SPLIT|<\/?UNTRUSTED_[A-Z_]+>", Options),
            PromptInjectionRisk.Medium),
    ];

    private static readonly string[] ForbiddenMarkers =
    [
        "проверка пропущена",
        "пропускаю проверку",
        "по просьбе студента",
        "ставлю 10 по просьбе",
        "skip the check",
        "as requested by the student",
        "give full credit",
    ];

    public PromptInjectionAnalysis Analyze(string? text)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return PromptInjectionAnalysis.Empty();
        }

        var indicators = new List<string>();
        var fragments = new List<string>();
        var riskLevel = PromptInjectionRisk.None;

        foreach (var pattern in _patterns)
        {
            var match = pattern.Regex.Match(text);

            if (!match.Success)
            {
                continue;
            }

            indicators.Add(pattern.Indicator);
            fragments.Add(ExtractFragment(text, match.Index, match.Length));
            riskLevel = PromptInjectionRisk.Max(riskLevel, pattern.RiskLevel);
        }

        return new PromptInjectionAnalysis
        {
            Detected = indicators.Count > 0,
            RiskLevel = riskLevel,
            Indicators = Unique(indicators),
            Fragments = Unique(fragments),
        };
    }

    public T MergeResultFields<T>(T result, PromptInjectionAnalysis analysis) where T : IPromptInjectionFields
    {
        var detected = result.PromptInjectionDetected || analysis.Detected;
        var risk = PromptInjectionRisk.Max(result.PromptInjectionRisk ?? PromptInjectionRisk.None, analysis.RiskLevel);
        var fragments = Unique([.. result.PromptInjectionFragments ?? [], .. analysis.Fragments]);
        var securityComment = !string.IsNullOrEmpty(result.SecurityComment)
            ? result.SecurityComment
            : (detected ? "Potential prompt injection was detected in untrusted report content." : string.Empty);

        result.PromptInjectionDetected = detected;
        result.PromptInjectionRisk = risk;
        result.PromptInjectionFragments = [.. fragments];
        result.SecurityComment = securityComment;

        return result;
    }

    public void AssertGeneratedReviewAllowed(string? review, IEnumerable<string>? advantages, IEnumerable<string>? disadvantages)
    {
        var parts = new List<string> { review ?? string.Empty };
        parts.AddRange(advantages ?? []);
        parts.AddRange(disadvantages ?? []);

        var text = string.Join("\n", parts).ToLowerInvariant();

        if (ForbiddenMarkers.Any(marker => text.Contains(marker, StringComparison.Ordinal)))
        {
            throw new InvalidOperationException("Model response appears to follow a prompt injection instruction");
        }
    }

    public string FormatSecurityContext(PromptInjectionAnalysis analysis)
    {
        var payload = new
        {
            promptInjectionDetected = analysis.Detected,
            promptInjectionRisk = analysis.RiskLevel,
            indicators = analysis.Indicators,
            suspiciousFragments = analysis.Fragments,
        };

        return JsonSerializer.Serialize(payload, new JsonSerializerOptions
        {
            WriteIndented = true,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        });
    }

    private static string ExtractFragment(string text, int index, int length)
    {
        const int padding = 80;
        var start = Math.Max(0, index - padding);
        var end = Math.Min(text.Length, index + length + padding);
        var slice = text[start..end];

        return Regex.Replace(slice, @"\s+", " ").Trim();
    }

    private static List<string> Unique(IEnumerable<string> items)
    {
        var seen = new HashSet<string>();
        var result = new List<string>();

        foreach (var item in items)
        {
            if (!string.IsNullOrEmpty(item) && seen.Add(item))
            {
                result.Add(item);
            }
        }

        return result;
    }
}
